R/loadFromMothur.R

```diff
@@ -0,0 +1,206 @@
+      taxa_tab <- S4Vectors:::make_zero_col_DataFrame(nrow(feature_tab))
+    }
+    
+    # If colData information exists, gets that. Otherwise, sample_tab is just data frame without information
+    if (!is.null(sampleMetaFile)) {
+      sample_meta <- .read_mothur_sample_meta(sampleMetaFile, feature_tab)
+    } else {
+      sample_meta <- S4Vectors:::make_zero_col_DataFrame(ncol(feature_tab))
+    }
+    
+    if (!is.null(phyTreeFile)) {
+      tree <- .read_qza(phyTreeFile, ...)
+    } else {
+      tree <- NULL
+    }
+    
+    # if row.names(feature_tab) is a DNA sequence,  set it as refseq
+    if (!is.null(refSeqFile)){
+      refseq <- .read_qza(refSeqFile, ...)
+    } else if (featureNamesAsRefSeq) {
+      refseq <- .rownames_as_dna_seq(rownames(feature_tab))
+    } else {
+      refseq <- NULL
+    }
+    
+    return(TreeSummarizedExperiment(
+      assays = S4Vectors::SimpleList(counts = feature_tab),
+      rowData = taxa_tab,
+      colData = sample_meta,
+      rowTree = tree,
+      referenceSeq = refseq
+    ))
+    
+}
+
+.read_mothur_feature <- function(featureTableFile, ...){
+
+    # Reads the file
+    assay <- read.table(featureTableFile, check.names=FALSE, header=TRUE,
+                      sep="\t", stringsAsFactors=FALSE)
+  
+    # File contains additional columns "label", "numOtus", and "Group". They are removed
+    assay$label <- NULL
+    assay$numOtus <- NULL
+    # Information of "Group" column is saved, it includes taxa information
+    x <- assay$Group
+    assay$Group <- NULL
+    
+    # Initializes rownames
+    rownames(assay) <- NULL
+    # Saves taxa to rownames
+    rownames(assay) <- x
+    
+    # Transposes and converts assay from dataframe to matrix
+    assay <- t(as.matrix(assay))
+    
+    return(assay)
+    
+}
+
+.read_mothur_taxonomy <- function(taxonomyTableFile, feature_tab, ...){
+    
+    # Reads the file
+    rowData <- read.table(taxonomyTableFile, check.names=FALSE,
+                          header=TRUE, sep="\t", stringsAsFactors=FALSE)
+    # Deletes additional information between taxa levels.
+    rowData$Taxonomy <- gsub("[\"]", "", rowData$Taxonomy)
+    rowData$Taxonomy <- gsub("[(1-100)]", "", rowData$Taxonomy)
+    # Separate taxa to own columns
+    rowData <- tidyr::separate(rowData, "Taxonomy", into=c("Kingdom",
+                                                           "Phylum", "Order", "Class", "Family", "Genus"), sep=";",
+                               extra="merge")
+    # Deletes ";" from the end of Genus names
+    rowData$Genus <- gsub(";", "", rowData$Genus)
+    # Deletes addition "Size" column
+    rowData$Size <- NULL
+    
+    # Adds taxa to rownames
+    rownames(rowData) <- rowData$OTU
+    # Deletes addition "OTU" column
+    rowData$OTU <- NULL
+    
+    # Creates a matrix from the table
+    rowData <- as.matrix(rowData)
+    
+    # Gets only those taxa that are included in feature_tab
+    rowData <- rowData[rownames(feature_tab),]
```

Comment:
Why? Additional data is allowed isn't it? Why not import it as well?

And also, why is the subset necessary? Shouldn't the dimension of the data match out of the box?

Reply:
rownames should match, and subsetting is unnecessary. I checked my data, and for some reason in taxonomic file random taxa are dropped from the middle

Reply:
They should match, but if someone mixes up their files? A lot of `NA` get produced or a subset of bound error is triggered, without any clear indication, what the problem is.

If the rownames should match, than you have to make sure.